core/src/main/java/org/elasticsearch/index/seqno/LocalCheckpointService.java

```diff
@@ -0,0 +1,149 @@
+
+    /**
+     * we keep a bit for each seq No that is still pending. to optimize allocation, we do so in multiple arrays
+     * allocating them on demand and cleaning up while completed. This setting controls the size of the arrays
+     */
+    public static String SETTINGS_BIT_ARRAYS_SIZE = "index.seq_no.checkpoint.bit_arrays_size";
+
+    /** default value for {@link #SETTINGS_BIT_ARRAYS_SIZE} */
+    final static int DEFAULT_BIT_ARRAYS_SIZE = 1024;
+
+
+    /**
+     * an order list of bit arrays representing pending seq nos. The list is "anchored" in {@link #firstSeqNoInProcessSeqNo}
+     * which marks the seqNo the fist bit in the first array corresponds to.
+     */
+    final LinkedList<FixedBitSet> processedSeqNo;
+    final int bitArraysSize;
+    long firstSeqNoInProcessSeqNo = 0;
+
+    /** the current local checkpoint, i.e., all seqNo lower&lt;= this number have been completed */
+    volatile long checkpoint = -1;
+
+    /** the next available seqNo - used for seqNo generation */
+    volatile long nextSeqNo = 0;
+
+
+    public LocalCheckpointService(ShardId shardId, IndexSettings indexSettings) {
+        super(shardId, indexSettings);
+        bitArraysSize = indexSettings.getSettings().getAsInt(SETTINGS_BIT_ARRAYS_SIZE, DEFAULT_BIT_ARRAYS_SIZE);
+        if (bitArraysSize <= 0) {
+            throw new IllegalArgumentException("[" + SETTINGS_BIT_ARRAYS_SIZE + "] must be positive. got [" + bitArraysSize + "]");
+        }
+        processedSeqNo = new LinkedList<>();
+    }
+
+    /**
+     * issue the next sequence number
+     **/
+    public synchronized long generateSeqNo() {
+        return nextSeqNo++;
+    }
+
+    /**
+     * marks the processing of the given seqNo have been completed
+     **/
+    public synchronized void markSeqNoAsCompleted(long seqNo) {
+        // make sure we track highest seen seqNo
+        if (seqNo >= nextSeqNo) {
+            nextSeqNo = seqNo + 1;
+        }
+        if (seqNo <= checkpoint) {
+            // this is possible during recovery where we might replay an op that was also replicated
+            return;
+        }
+        FixedBitSet bitSet = getBitSetForSeqNo(seqNo);
+        int offset = seqNoToBitSetOffset(seqNo);
+        bitSet.set(offset);
+        if (seqNo == checkpoint + 1) {
+            updateCheckpoint();
+        }
+    }
+
+    /** get's the current check point */
+    public long getCheckpoint() {
+        return checkpoint;
+    }
+
+    /** get's the maximum seqno seen so far */
+    public long getMaxSeqNo() {
+        return nextSeqNo - 1;
+    }
+
+    /**
+     * moves the checkpoint to the last consecutively processed seqNo
+     * Note: this method assumes that the seqNo following the current checkpoint is processed.
+     */
+    private void updateCheckpoint() {
+        assert Thread.holdsLock(this);
+        assert checkpoint - firstSeqNoInProcessSeqNo < bitArraysSize : "checkpoint to firstSeqNoInProcessSeqNo is larger then a bit set";
```

Comment:
Is `checkpoint - firstSeqNoInProcessSeqNo + 1 <= bitArraySize` clearer (maybe since `checkpoint == firstSeqNoInProcessSeqNo + bitArraySize - 1` appears later)?